src/Orleans.Core/Messaging/ClientMessageCenter.cs

```diff
@@ -293,32 +308,28 @@ private ValueTask<Connection> GetGatewayConnection(Message msg)
             var gatewayConnection = this.connectionManager.GetConnection(addr);
             if (gatewayConnection.IsCompletedSuccessfully)
             {
-                var result = this.TryUpdateConnection(index, gatewayConnection.Result, weakRef);
-                if (result == null) return this.GetGatewayConnection(msg);
-                return new ValueTask<Connection>(result);
+                this.UpdateBucket(index, gatewayConnection.Result);
+                return gatewayConnection;
             }
```

Comment:
The other changes in this file are about simplifying the logic. We have a single lock for modifying the buckets & WeakRefs and we re-use the WeakRefs, updating them instead of replacing them. Hopefully it's easier to understand than before.